### PR TITLE
[CP] Support breakable CUDA graphs for zigzag strategy

### PR DESCRIPTION
**Pull request by Fridge003**
## Motivation

Split PR 2 of #32714. Enable breakable prefill CUDA graphs for the validated zigzag context-parallel path as an independent PR, then incorporate the requested review refactor.

## Modifications

- allow full zigzag prefill CP with TensorRT-LLM MHA to retain breakable prefill CUDA graphs, without an architecture restriction
- keep CP-specific BCG eligibility, static inputs, capture preparation, replay, gather, and eager logits-tail handling in `sglang.srt.layers.cp.bcg`
- capture only token buckets where zigzag CP is valid
- gate live batches at the prefill CUDA graph dispatch boundary so batches that cannot use CP-v2 fall back to eager execution
- extend `test_gpt_oss_4gpu_mxfp4_cp.py` with `--cuda-graph-backend-prefill breakable`
- remove the two standalone unit-test files requested in review

The final PR owns 5 files and has no changed-file overlap with the other split PRs.

## Accuracy Tests

- `baizhou-dev` (4x NVIDIA B200):
  - `PYTHONPATH=/scratch/sglang/python:$PYTHONPATH python3 test/registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py`
  - result: `Ran 1 test in 101.455s` / `OK`
  - GPT-OSS GPQA score: `0.6060606060606061` (required `0.58`)
  - all 4 ranks captured 73 valid prefill BCG token buckets
  - live requests logged 60 rank-level `Prefill batch ... cuda graph: True` entries, confirming replay was exercised
- local `python3 -m py_compile` passed for all touched Python files
- local pre-commit passed for all touched files

## Speed Tests and Profiling

No standalone benchmark was run. The required 4xB200 registered test reported `11993.292 token/s` output throughput during GPQA evaluation.

## Checklist

- [x] Format code with pre-commit.
- [x] Run the requested GPT-OSS prefill CP + breakable CUDA graph test on 4xB200.
- [x] Keep the three split PR file sets independent.
- [x] Follow the requested review comments without resolving review threads automatically.

---
### CI States

Latest PR Test (Base): :x: [Run #30785759856](https://github.com/sgl-project/sglang/actions/runs/30785759856)
Latest PR Test (Extra): :x: [Run #30785759783](https://github.com/sgl-project/sglang/actions/runs/30785759783)

### CODE REVIEW

File: test/registered/unit/server_args/test_flashinfer_a2a_cp.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
+import unittest
```
Comment (Fridge003) [inline test/registered/unit/server_args/test_flashinfer_a2a_cp.py]: Remove this test

File: test/registered/unit/model_executor/runner/test_prefill_cuda_graph_cp.py [hunk 2]
```diff
@@ -0,0 +1,513 @@
+import unittest
```
Comment (Fridge003) [inline test/registered/unit/model_executor/runner/test_prefill_cuda_graph_cp.py]: Remove this test

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -4446,6 +4446,18 @@ def _disable_tc_piecewise_cudagraph_if_incompatible(self):
             if predicate():
                 self.cuda_graph_config.prefill.backend = Backend.DISABLED
 
+    def _supports_breakable_prefill_cp(self) -> bool:
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Move this function to python/sglang/srt/layers/cp/bcg.py
You might pass server_args as input to this functions

File: python/sglang/srt/server_args.py [hunk 4]
```diff
@@ -4446,6 +4446,18 @@ def _disable_tc_piecewise_cudagraph_if_incompatible(self):
...
+        resolved = self._resolved()
+        prefill_attention_backend, _ = self._resolved_attention_backends()
+        return (
+            self.enable_prefill_cp
+            and resolved.attn_cp_size == self.tp_size
+            and self.cp_strategy == "zigzag"
+            and prefill_attention_backend == "trtllm_mha"
+            and self.get_model_config().hf_config.architectures == ["GptOssForCausalLM"]
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Remove this line of and self.get_model_config().hf_config.architectures == ["GptOssForCausalLM"]

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 5]
```diff
@@ -581,6 +614,133 @@ def _get_layer_model_positions(self, forward_batch: ForwardBatch) -> torch.Tenso
...
+        positions = self.cp_positions[:captured_local_tokens]
+        input_embeds.zero_()
+        positions.zero_()
+        input_embeds[:live_local_tokens].copy_(local_input_embeds)
+        positions[:live_local_tokens].copy_(local_positions)
+        forward_batch.input_embeds = input_embeds
+        forward_batch.positions = positions
+        return live_local_tokens
+
+    def _validate_cp_flashinfer_dispatch_capacity(
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Remove this function from this PR

File: test/registered/unit/server_args/test_flashinfer_a2a_cp.py (hunk 1 above)
Comment (Fridge003) [inline test/registered/unit/server_args/test_flashinfer_a2a_cp.py]: For testing this feature, we might enable breakable cuda graph in test/registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 6]
```diff
@@ -447,6 +458,28 @@ def __init__(self, model_runner: ModelRunner):
                     for name in _PREFILL_STATIC_FIELDS
                 }
 
+        server_args = model_runner.server_args
+        self.enable_cp_v2_body_capture = (
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Move  `enable_cp_v2_body_capture`  to python/sglang/srt/layers/cp/utils.py, and rename it as `enable_cp_v2_bcg_capture`

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 7]
```diff
@@ -350,6 +356,11 @@ def __init__(self, model_runner: ModelRunner):
         self._is_full_backend = False
         # Same ordering requirement: capture_prepare reads this.
         self._capture_lora = False
+        self.enable_cp_v2_body_capture = False
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Can we group these variables into a structure, maybe `PrefillCPBCGInput`, and put it under python/sglang/srt/layers/cp/bcg.py

So here we can set `self.prefill_cp_bcg_input = ....`

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 8]
```diff
@@ -53,6 +53,12 @@
 from sglang.srt.configs.model_config import is_deepseek_dsa
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py:53]: Open a new file python/sglang/srt/layers/cp/bcg.py dedicated for Prefill CP + BCG

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 9]
```diff
@@ -581,6 +614,133 @@ def _get_layer_model_positions(self, forward_batch: ForwardBatch) -> torch.Tenso
 
         return forward_batch.positions
 
+    def _prepare_cp_static_inputs(
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Move this function to python/sglang/srt/layers/cp/bcg.py

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 10]
```diff
@@ -1290,6 +1450,16 @@ def capture_one_shape(self, size: int, *, prefix_num_chunks: int = 0) -> None:
         """
         num_tokens = size
         forward_batch, attn_backend = self.capture_prepare(num_tokens)
+        if self.enable_cp_v2_body_capture:
+            required_local_tokens = self._prepare_cp_static_inputs(
+                forward_batch,
+                static_num_tokens=num_tokens,
+                capture=True,
+            )
+            self._validate_cp_flashinfer_dispatch_capacity(
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Remove the calling of self._validate_cp_flashinfer_dispatch_capacity

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 11]
```diff
@@ -1526,12 +1696,75 @@ def _slot(name):
...
+            metadata_forward_batch = static_forward_batch
+
         self._prepare_forward_metadata_for_replay(
-            forward_batch, static_forward_batch, static_num_tokens
+            metadata_forward_batch, static_forward_batch, static_num_tokens
         )
 
         return static_forward_batch
 
+    def _execute_cp_body_capture(
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Move this function to python/sglang/srt/layers/cp/bcg.py

File: python/sglang/srt/server_args.py [hunk 12]
```diff
@@ -6589,6 +6591,14 @@ def _validate_cutedsl_a2a_token_budget(self):
                 f"(e.g. --max-prefill-tokens) to <= {max_cutedsl_tokens}."
             )
 
+    def _supports_flashinfer_a2a_parallelism(self) -> bool:
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Remove this function

File: python/sglang/srt/server_args.py [hunk 13]
```diff
@@ -6669,8 +6679,8 @@ def _handle_a2a_moe(self):
             )
         if self.moe_a2a_backend == "flashinfer":
             assert (
-                resolved_view(self).enable_dp_attention and self.dp_size == self.tp_size
-            ), "Flashinfer MoE A2A is only supported with dp_size == tp_size and --enable-dp-attention"
+                self._supports_flashinfer_a2a_parallelism()
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Restore these two lines

File: test/registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py [hunk 14]
```diff
@@ -1,12 +1,24 @@
...
+
 from sglang.test.ci.ci_register import register_cuda_ci
 from sglang.test.gpt_oss_common import BaseTestGptOss
+from sglang.test.test_utils import DEFAULT_URL_FOR_TEST
 
 register_cuda_ci(est_time=220, stage="extra-b", runner_config="4-gpu-b200")
 
 
 class TestGptOss4GpuMxfp4CP(BaseTestGptOss):
+    def _check_streaming_responses_api_request(self, model):
```
Comment (Fridge003) [inline test/registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py]: Remove this _check_streaming_responses_api_request

File: python/sglang/srt/layers/cp/utils.py [hunk 15]
```diff
@@ -135,6 +135,16 @@ def enable_cp_v2() -> bool:
     return bool(envs.SGLANG_ENABLE_CP_V2.get())
 
 
+def enable_cp_v2_bcg_capture(server_args) -> bool:
```
Comment (Fridge003) [inline python/sglang/srt/layers/cp/utils.py]: Move this to ./bcg.py

File: python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py [hunk 16]
```diff
@@ -1068,6 +1091,9 @@ def can_replay_locally(
         return True
 
     def can_run_graph(self, forward_batch: ForwardBatch) -> bool:
+        if self.enable_cp_v2_bcg_capture and not is_cp_v2_active(forward_batch):
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/prefill_cuda_graph_runner.py]: Remove this two lines, since not is_cp_v2_active(forward_batch) has been included in self.enable_cp_v2_bcg_capture

**Comment by Fridge003** [comment]
/rerun-test test/registered/cp/test_cp_strategy_unit.py test/registered/cp/test_gqa_prefill_cp.py test/registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py  test/registered/cuda_graph/breakable/test_breakable_cuda_graph.py  test/registered/cuda_graph/breakable/test_bcg_with_speculative_decoding.py
